backend/app/treasury/tests/test_covalent.py

```diff
@@ -0,0 +1,327 @@
+        treasury_transfers = await covalent._get_transfer_data("", "", 1)
+
+        assert not treasury_transfers
+
+
+@mark.asyncio
+async def test_covalent_transfers_bad_transfers(monkeypatch: MonkeyPatch):
+    monkeypatch.setattr(AsyncClient, "get", return_mocked_resp)
+    _covalent_transfers_v2_transfers = deepcopy(covalent_transfers_v2_transfers)
+    for item in _covalent_transfers_v2_transfers:
+        del item["transfers"]
+
+    with raises(KeyError):
+        tokens = list(
+            covalent._gen_transfers(
+                {
+                    "address": "0x6d6f636b65645f7472656173757279",
+                    "quote_currency": "USD",
+                    "chain_id": 1,
+                    "items": _covalent_transfers_v2_transfers,
+                }
+            )
+        )
+
+        assert not tokens
+
+    _covalent_transfers_v2_transfers = deepcopy(covalent_transfers_v2_transfers)
+    for item in _covalent_transfers_v2_transfers:
+        del item["transfers"][0]["contract_decimals"]
+
+    with raises(KeyError):
+        tokens = list(
+            covalent._gen_transfers(
+                {
+                    "address": "0x6d6f636b65645f7472656173757279",
+                    "quote_currency": "USD",
+                    "chain_id": 1,
+                    "items": _covalent_transfers_v2_transfers,
+                }
+            )
+        )
+
+        assert not tokens
+
+
+@mark.asyncio
+async def test_covalent_get_token_transfers(monkeypatch: MonkeyPatch):
+    fake_provider = FakeRedis()
+    monkeypatch.setattr(AsyncClient, "get", return_mocked_resp)
+    monkeypatch.setattr("backend.app.treasury.adapters.covalent.db", fake_provider)
+    monkeypatch.setattr(
+        MockResponse,
+        "json",
+        lambda *_: {
+            "data": {
+                "address": "0x6d6f636b65645f7472656173757279",
+                "quote_currency": "USD",
+                "chain_id": 1,
+                "items": covalent_transfers_v2_transfers,
+            }
+        },
+    )
+
+    sanitized_token_transfers = await covalent.get_token_transfers("0xa", "0xb", 1)
+    exp_sanitized_token_transfers = [
+        covalent.CovalentTransfer(
+            timestamp=parse(
+                covalent_transfers_v2_transfers[1]["transfers"][0]["block_signed_at"]
+            ),
+            balance=int(covalent_transfers_v2_transfers[1]["transfers"][0]["delta"])
+            / 1e18,
+        ),
+        covalent.CovalentTransfer(
+            timestamp=parse(
+                covalent_transfers_v2_transfers[0]["transfers"][0]["block_signed_at"]
+            ),
+            balance=int(covalent_transfers_v2_transfers[0]["transfers"][0]["delta"])
+            / 1e18
+            + int(covalent_transfers_v2_transfers[1]["transfers"][0]["delta"]) / 1e18,
+        ),
+    ]
+
+    assert sanitized_token_transfers == exp_sanitized_token_transfers
+
+    fake_cache_key = "{treasury_address}_{contract_address}_{chain_id}_{date}"
```

Comment:
You should split this test in two, starting from this line to keep things unitary. There is going to be repeated code at start of the two test functions, but that's fine (and if it ever gets bigger, pytest has tools to help us with that).

Also, here in the second part of the test, you are testing both the behavior of the `get_token_transfers` function and of the Redis library all at once. That's totally fine for now, because writing extra mocks for the calls to `db.hset` and such would be a bit overkill.  
But as a side note, that hints at that we should probably start thinking about an intermediary cache layer/middleware, rather than directly calling `db.hset`, so as to make `get_token_transfers` more agnostic of the cache mechanism.